incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -31,44 +39,182 @@ type HierarchicalNamespaceReconciler struct {
 
 	forest *forest.Forest
 	hcr    *HierarchyConfigReconciler
+
+	// Affected is a channel of event.GenericEvent (see "Watching Channels" in
+	// https://book-v1.book.kubebuilder.io/beyond_basics/controller_watches.html) that is used to
+	// enqueue additional objects that need updating.
+	Affected chan event.GenericEvent
 }
 
 // Reconcile sets up some basic variables and then calls the business logic. It currently
 // only handles the creation of the namespaces but no deletion or state reporting yet.
 func (r *HierarchicalNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// TODO report error state if the webhook is bypassed - see issue
-	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/459
-	if ex[req.Namespace] {
-		return ctrl.Result{}, nil
-	}
-
+	ctx := context.Background()
 	log := r.Log.WithValues("trigger", req.NamespacedName)
 	log.Info("Reconciling HNS")
 
-	// Names of the hierarchical namespace and the current namespace.
+	// Get names of the hierarchical namespace and the current namespace.
 	nm := req.Name
 	pnm := req.Namespace
 
-	// Set RequiredChildOf (Owner) in the forest of the hierarchical namespace to the
-	// current namespace.
+	// Get instance from apiserver. If the instance doesn't exist, we don't want to reconcile
+	// it since it may trigger the HC reconciler to recreate the namespace that was just deleted.
```

Comment:
Looks from the code, if instance does not exist and we return an error instead of nil, `Reconcile()` of `HierarchicalNamespaceReconciler` (this method) will be called again, which will get the instance, and we are trapped in an infinite loop. Why it will trigger `HC reconciler to recreate the namespace that was just deleted.`?  Feel free to correct me if I miss anything. :)

Reply:
Ah, line 67-70 did return a `nil` to prevent a retry.

It "may" trigger the HC reconciler to recreate the namespace is the namespace is missing, see `syncMissing`.

Reply:
Thanks so much for the explanation offline.

Reply:
You are welcome and thanks for bring this up!